realm/realm-library/src/main/java/io/realm/internal/android/JsonUtils.java

```diff
@@ -41,15 +43,18 @@ public static Date stringToDate(String date) {
         if (matcher.find()) {
             String dateMatch = matcher.group(1);
             return new Date(Long.parseLong(dateMatch));
-        } else {
-            return new Date(Long.parseLong(date));
         }
+        try {
+            return ISO8601Utils.parse(date, new ParsePosition(0));
+        } catch (ParseException ignored) {
+        }
+        return new Date(Long.parseLong(date));
```

Comment:
I'm a bit worried about having this last. It means that if you actually have this format, you have to deal with a `ParseException` for each object, which is very expensive. Is there a lightweight way of detecting this scenario? A simple regexp would do I assume? `/-?\d*/`. Perhaps `string.contains("-")` or some other char that is always present (not sure there is any though judging from the unit tests).